lib/Basics/expected.h

```diff
@@ -0,0 +1,217 @@
+#pragma once
+#include <stdexcept>
+#include <type_traits>
+#include "Assertions/Assert.h"
+
+namespace arangodb {
+template<typename T>
+struct expected {
+  expected() noexcept {}
+
+  template<typename... Args>
+  explicit expected(std::in_place_t, Args&&... args) noexcept(
+      std::is_nothrow_constructible_v<T, Args...>) requires
+      std::constructible_from<T, Args...> {
+    new (&_value) T(std::forward<Args>(args)...);
+    _state = kValue;
+  }
+
+  explicit expected(std::exception_ptr ex) noexcept
+      : _exception(std::move(ex)), _state(kException) {
+    static_assert(std::is_nothrow_move_constructible_v<std::exception_ptr>);
+  }
+
+  expected(expected const& other) noexcept(
+      std::is_nothrow_copy_constructible_v<T>) requires
+      std::is_copy_constructible_v<T> {
+    copy_from(other);
+  }
+
+  expected& operator=(expected const& other) noexcept(
+      std::is_nothrow_copy_constructible_v<T>&&
+          std::is_nothrow_destructible_v<T>&& std::is_nothrow_copy_assignable_v<
+              T>) requires(std::is_copy_constructible_v<T>&&
+                               std::is_copy_assignable_v<T>) {
+    if (this != &other) {
+      if (other._state == kValue && _state == kValue) {
+        _value = other._value;
+      } else {
+        reset();
+        copy_from(other);
+      }
+    }
+    return *this;
+  }
+
+  expected(expected&& other) noexcept(
+      std::is_nothrow_move_constructible_v<T>) requires
+      std::move_constructible<T> {
+    move_from(std::move(other));
+  }
+
+  expected& operator=(expected&& other) noexcept(
+      std::is_nothrow_move_constructible_v<T>&&
+          std::is_nothrow_destructible_v<T>&& std::is_nothrow_move_assignable_v<
+              T>) requires(std::move_constructible<T>&&
+                               std::is_move_assignable_v<T>) {
```

Comment:
I think we should remove the `is_nothrow_move_constructible_v` and `move_constructible` checks here. I know that strictley speaking we need the for the `move_from` call, but that's why `move_from` has theses checks, so I wouldn't repeat them here.
```suggestion
  expected& operator=(expected&& other) noexcept(
        std::is_nothrow_destructible_v<T>&& std::is_nothrow_move_assignable_v<
            T>) requires(std::is_move_assignable_v<T>) {
```